test/apps/iodemo/io_demo.cc

```diff
@@ -99,21 +137,30 @@ class P2pDemoCommon : public UcxContext {
     /* Asynchronous IO message */
     class IoMessage : public UcxCallback {
     public:
-        IoMessage(size_t buffer_size, io_op_t op, uint32_t sn, size_t data_size) :
-            _buffer(malloc(buffer_size)) {
+        IoMessage() : _buffer(NULL), _pool(NULL) {
+        }
+        
+        void init(size_t buffer_size, io_op_t op, uint32_t sn,
+                    size_t data_size, MemoryPool<IoMessage>* pool) {
+            if (_buffer == NULL) {
+                _buffer = malloc(buffer_size); 
+            }
             iomsg_hdr_t *hdr = reinterpret_cast<iomsg_hdr_t*>(_buffer);
             assert(sizeof(*hdr) <= buffer_size);
             hdr->op          = op;
             hdr->sn          = sn;
             hdr->data_size   = data_size;
+            _pool            = pool;
         }
 
         ~IoMessage() {
             free(_buffer);
         }
 
         virtual void operator()(ucs_status_t status) {
-            delete this;
+            if (_pool) {
```

Comment:
can we pass the pool in constructor instead of init(), then it will never be NULL?

Reply:
About this comment and above one "can we allocate the buffer in constructor?", in my opinnion, if we want to do these things in contractors, we have to put the parameters to contructor functions, for example buffer_size, pool pointer. IoMessage as a component of memory pool, the contractor is executed in MemoryPool::get(). Then we need to deal with these parameters in MemoryPool. But for different kind of component will need different set of parameters, so we don't want to deal with this in MemoryPool but leave it to the component class itself. So put the parameters to init() function and do the things in init and make the contractor without any parameters.

Reply:
i would pass buffer size and memory pool in constructor, directly from MemoryPool::get()
(for cases where T does not have buffer, we can pass 0)
this way it's more safe - we know the _buffer size is the same for every object in the memory pool

Reply:
@yosefe Then we can't make MemoryPool common. For exaplem, sometimes we don't only want to allocate one buffer for one item, or we need to specify some other things for each item besides buffer_size and memory pool pointer.
But since we just use MemoryPool in io demo, we don't need to cover so many cases won't occur in io demo, so I also agree that it's fine to change it to constructor. It has the benefits you mentioned.
In my fix I added the buffer size as the parameter of MemoryPool::get(). Actually the safer way could be passing it as the parameter of contractor MemoryPool::MemoryPool(). But this will make the change more complex since sometimes it's hard to know how big buffer size should be allocated at the place we create the Memory pool object. If you think it's necessary to change to pass the buffer size through parameter MemoryPool::MemoryPool() or have other good suggestions, please feel free to let me know. Thanks!

Reply:
@leibin2014 i think we need to pass buffer size to MemoryPool constructor, since the (maximal) buffer size is a user global parameter, we should be able to know it.

Reply:
@yosefe Fixed. Thanks!